coordinator/src/main/scala/filodb.coordinator/queryengine2/FailureProvider.scala

```diff
@@ -0,0 +1,194 @@
+        }
+      }
+  }
+
+  def plan(lp: LogicalPlan, failures: Seq[FailureTimeRange]): Seq[Route] = {
+
+    if (!isPeriodicSeriesPlan(lp) || !hasSingleTimeRange(lp))
+      new LocalRoute(None)
+
+    val nonOverlappingFailures = removeLargerOverlappingFailures(failures)
+    val time = getTimeFromLogicalPlan(lp)
+    if(nonOverlappingFailures.last.timeRange.startInMillis < time.startInMillis )
+      return Seq(LocalRoute(None))  // No failure in this time range
+    logger.info("Logical plan time:" + time)
+
+    splitQueryTime(nonOverlappingFailures, 0, time.startInMillis, time.endInMillis)
+
+  }
+
+  def splitQueryTime(failure: Seq[FailureTimeRange], index: Int, start: Long, end: Long): Seq[Route] = {
+
+    if (index >= failure.length)
+      return Nil
+    // traverse query range time from left to right , break at failure start
+    var i = index + 1
+
+    failure(index).dispatcher.map { x =>
+      // Handle local failure
+      // Traverse till we get a remote failure
+      while ((i < failure.length) && (failure(i).dispatcher.isDefined))
+        i = i + 1
+
+      if (i < failure.length)
+        RemoteRoute(Some(TimeRange(start, failure(i).timeRange.startInMillis)),
+          x) +: splitQueryTime(failure, i, failure(i).timeRange.startInMillis, end)
+      else
+      // Last failure so no further splitting required
+        Seq(RemoteRoute(Some(TimeRange(start, end)), x))
+
+    }.getOrElse {
+      // Iterate till we get a local failure
+      while ((i < failure.length) && (!failure(i).dispatcher.isDefined))
+        i = i + 1
+      if (i < failure.length) //need further splitting
+        LocalRoute(Some(TimeRange(start, failure(i).timeRange.startInMillis))) +:
+          splitQueryTime(failure, i, failure(i).timeRange.startInMillis, end)
+      else
+        Seq(LocalRoute(Some(TimeRange(start, end))))
+    }
+  }
+
+  def isPeriodicSeriesPlan(logicalPlan: LogicalPlan): Boolean = {
+    if (logicalPlan.isInstanceOf[RawSeriesPlan] || logicalPlan.isInstanceOf[MetadataQueryPlan])
+      return false;
+    true
+  }
+
+  def hasSingleTimeRange(logicalPlan: LogicalPlan): Boolean = {
+    if (logicalPlan.isInstanceOf[BinaryJoin]) {
+      val binaryJoin = logicalPlan.asInstanceOf[BinaryJoin]
+      val lhsTime = getTimeFromLogicalPlan(binaryJoin.lhs)
+      val rhsTime = getTimeFromLogicalPlan(binaryJoin.rhs)
+      return (lhsTime.startInMillis == rhsTime.startInMillis) && (lhsTime.endInMillis == rhsTime.endInMillis)
+    }
+    return true
+  }
+
+  def getTimeFromLogicalPlan(logicalPlan: LogicalPlan): TimeRange = {
+    logicalPlan match {
+      case lp: PeriodicSeries => val periodicSeries = lp.asInstanceOf[PeriodicSeries]
+        TimeRange(periodicSeries.start, periodicSeries.end)
+      case lp: PeriodicSeriesWithWindowing => val periodicSeriesWithWindowing = lp.
+        asInstanceOf[PeriodicSeriesWithWindowing]
+        TimeRange(periodicSeriesWithWindowing.start,
+          periodicSeriesWithWindowing.end)
```

Comment:
We need to discuss because there is an interesting twist to the plot here, and I am not sure if you recognized it. The range returned by RoutingPlanner will be based on start/step/end, but the routing algorithm should take into consideration the time range provided by RawSeries.rangeSelector.

Reply:
Changed according to look-back window.